src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -160,7 +168,41 @@ static IContainSagaData DeepClone(IContainSagaData source)
 
             public IContainSagaData GetSagaCopy()
             {
-                return DeepClone(data);
+                var canBeClonedShallowly = cloneMethods.GetOrAdd(data.GetType(), CanBeMemberwiselyCloned);
+                return canBeClonedShallowly ? memberwiseClone(data) : DeepClone(data);
+            }
+
+            static bool CanBeMemberwiselyCloned(Type type)
+            {
+                foreach (var fi in type.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy))
```

Comment:
Ok so this makes sure we get the whole hierarchy even if we have deeply inherited saga data